src/lsm/lsm_cursor.c

```diff
@@ -902,6 +902,104 @@ err:	__clsm_leave(clsm);
 }
 
 /*
+ * __clsm_random_chunk --
+ *	Pick a chunk at random, weighted by the size of all chunks. Weighting
+ * proportional to documents avoids biasing towards small chunks.
+ */
+static int
+__clsm_random_chunk(WT_SESSION_IMPL *session,
+    WT_CURSOR_LSM *clsm, WT_LSM_CHUNK **random_chunk)
+{
+	WT_DECL_RET;
+	WT_LSM_TREE *lsm_tree;
+	uint64_t checked_docs, i, rand_doc, total_docs;
+
+	lsm_tree = clsm->lsm_tree;
+	rand_doc = __wt_random(&session->rnd);
+	__wt_lsm_tree_readlock(session, lsm_tree);
+	for (total_docs = i = 0; i < clsm->nchunks; i++) {
+		total_docs += lsm_tree->chunk[i]->count;
+	}
+
+	rand_doc %= total_docs;
+
+	for (checked_docs = i = 0; i < clsm->nchunks; i++) {
+		checked_docs += lsm_tree->chunk[i]->count;
+		if (rand_doc <= checked_docs) {
+			*random_chunk = lsm_tree->chunk[i];
+			break;
+		}
+	}
+	__wt_lsm_tree_readunlock(session, lsm_tree);
+	return (ret);
+}
+
+/*
+ * __clsm_next_random --
+ *	WT_CURSOR->next method for the LSM cursor type when configured with
+ * next_random.
+ */
+static int
+__clsm_next_random(WT_CURSOR *cursor)
+{
+	WT_CURSOR_LSM *clsm;
+	WT_CURSOR *c;
+	WT_DECL_RET;
+	WT_LSM_CHUNK *random_chunk;
+	WT_SESSION_IMPL *session;
+	const char *cfg[3];
+
+	clsm = (WT_CURSOR_LSM *)cursor;
+	random_chunk = NULL;
+
+	CURSOR_API_CALL(cursor, session, next, NULL);
+	WT_CURSOR_NOVALUE(cursor);
+	WT_ERR(__clsm_enter(clsm, false, false));
+	cfg[0] = WT_CONFIG_BASE(session, WT_SESSION_open_cursor);
+	cfg[1] = "next_random=true";
+	cfg[2] = NULL;
+
+	do {
+		WT_ERR(__clsm_random_chunk(session, clsm, &random_chunk));
+		/*
+		 * Open a cursor on the chunk and pick a random key from within.
+		 */
+		WT_ERR(__wt_open_cursor(
+		    session, random_chunk->uri, NULL, cfg, &c));
+		ret = c->next(c);
+		/*
+		 * Sometimes we may run on an empty chunk or for some reason be
+		 * unable to get a random doc. Thanks okay and we retry.
+		 */
+		if (ret == WT_NOTFOUND)
+			continue;
+
+		WT_ERR(ret);
+		WT_ERR(c->get_key(c, &cursor->key));
+		WT_ERR(c->close(c));
+
+		/*
+		 * Search back through the tree from the top to resolve any
+		 * updates or tombstones. If we find a tombstone, then pick a
+		 * new random doc and try again. We retry forever as eventually
+		 * we should see a current doc or the tombstones will get
+		 * merged out.
+		 */
+		ret = __clsm_lookup(clsm, &cursor->value);
```

Comment:
In case __clsm_lookup fails for reasons other than WT_NOTFOUND, we return that error to the user. Even though __clsm_next_random() failed at that point, we end up modifying cursor->key. Ideally we shouldn't modify the passed arguments if our call failed.